### DISCUSSION TITLE
How to generate multiple exports in package.json using @nrwl/web:rollup?

### DISCUSSION THREAD

**Discussion by runoncedev** [nrwl/nx#11502] [General]
I'm trying to build a package with `@nrwl/web:rollup` that uses [`package.json` `exports`](https://nodejs.org/api/packages.html#package-entry-points) to expose multiple entry points:

```
{
  "exports": {
    ".": {
      "types": "./index.d.ts",
      "default": "./index.js"
    },
    "./hooks": {
      "types": "./hooks/index.d.ts",
      "default": "./hooks/index.js"
    }
  }
}
```

I've extended the default Rollup.js config to include the extra `input` entry:

```
const nrwlConfig = require('@nrwl/react/plugins/bundle-rollup');

module.exports = (config) => {
  nrwlConfig(config);
  return {
    ...config,
    input: [config.input, 'packages/component-library/src/hooks'],
  };
};
```

And wired it up in the `project.json` file alognside the `generateExportsField` oprtion:

```
{
  "build": {
    "executor": "@nrwl/web:rollup",
    "outputs": ["{options.outputPath}"],
    "options": {
      "rollupConfig": "rollup.config.js",
      "generateExportsField": true
    }
  }
}
```
But after building my project I'm only getting the root export in the generated `package.json`:

```
{
  "exports": {
    ".": {
      "types": "./index.d.ts",
      "import": "./index.js"
    }
  }
}
```

Is there any way to tell Nx to generate exports for all input files?

**Comment by runoncedev** [comment]
The solution was simpler than I imagined.

I just needed to add my extra `exports` in the projects `package.json`:

```
{
  "exports": {
    "./hooks": {
      "types": "./hooks/index.d.ts",
      "default": "./hooks/index.js"
    }
  }
}
```

Nx will then merge this in the generated `package.json` during the build process.

**Reply by Geczy** [reply]
this is what worked for me too, thank you !

```js
  "exports": {
    "./dist/mongo/index.js": {
      "types": "./dist/mongo/index.d.ts",
      "default": "./dist/mongo/index.js"
    },
    "./dist/psql/index.js": {
      "types": "./dist/psql/index.d.ts",
      "default": "./dist/psql/index.js"
    }
  },
```

```js
import { PrismaClient as PrismaMongo } from '@dotabod/prisma/dist/mongo/index.js'
import { PrismaClient as PrismaPsql } from '@dotabod/prisma/dist/psql/index.js'
```

**Comment by DannyvanHolten** [comment]
I am getting the following error when I do this:

```
Error during bundle: Error: Invalid value for option "output.inlineDynamicImports" - multiple inputs are not supported when "output.inlineDynamicImports" is true.
```

**Reply by DannyvanHolten** [reply]
Apparently in my Nx config it switches to inlineDynamicImports: true when I define multiple inputs. Seems very weird to me. It also switches to umd mode.

I've "reset" the config to the following to get it work:

```
module.exports = (nxConfig) => {
  nrwlConfig(nxConfig);

  return {
    ...nxConfig,
    input: [nxConfig.input, 'packages/component-library/src/hooks'],
    output: {
      ...nxConfig.output,
      format: 'esm',
      entryFileNames: '[name].esm.js',
      chunkFileNames: '[name].esm.js',
      inlineDynamicImports: false,
    },
  };
};
```

**Comment by DannyvanHolten** [comment]
My next problem is that Nx is not using the bundled version but a symlink. So when I import this in remix inside the Nx repository. It doesn't recognize there is a server.esm.js

**Reply by MathurAditya724** [reply]
I also ran into this same problem yesterday and to resolve this import problem, I just added the package's name and its path in the root `tsconfig` file. But still, the build errors are there

```json
"paths": {
  "@mypackage/sample": ["packages/sample/src/index.ts"],
  "@mypackage/sample/validations": ["packages/sample/src/validations.ts"],
}
```
